clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -689,12 +689,20 @@ public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offs
      *         transactional.id is not authorized. See the exception for more details
      * @throws KafkaException if the producer has encountered a previous fatal or abortable error, or for any
      *         other unexpected error
+     * @throws TimeoutException if the time taken for committing the transaction has surpassed <code>max.block.ms</code>.
+     * @throws InterruptException if the thread is interrupted while blocked
      */
     public void commitTransaction() throws ProducerFencedException {
         throwIfNoTransactionManager();
         TransactionalRequestResult result = transactionManager.beginCommit();
         sender.wakeup();
-        result.await();
+        try {
+            if (!result.await(maxBlockTimeMs, TimeUnit.MILLISECONDS)) {
+                throw new TimeoutException("Timeout expired while committing transaction in " + maxBlockTimeMs + "ms.");
```

Comment:
This will probably take a little more work. If there is a timeout, the user should be able to retry. At the moment, I think this would just cause an illegal state error. We probably need to cache the `TransactionalRequestResult` somewhere so that we can continue on the next retry. We do something like this in `initTransactions()` already. It might be possible to factor out some common logic so that we don't just have a bunch of internal fields. Maybe we could move the caching into `TransactionManager` ∑here we would be able to validate the state.